internal/retry/config.go

```diff
@@ -0,0 +1,166 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/uber-go/mapdecode"
+	"go.uber.org/multierr"
+	iconfig "go.uber.org/yarpc/internal/config"
+	"go.uber.org/yarpc/x/config"
+)
+
+// PolicySpec defines how to construct a retry Policy.
+type PolicySpec struct {
+	// Retries indicates the number of retries will be attempted on
+	// failed requests.
+	Retries uint `config:"retries"`
+
+	// MaxRequestTimeout indicates the max timeout that every request
+	// through the retry middleware will have.  If the timeout is greater
+	// than the request timeout we'll use the request timeout instead.
+	MaxRequestTimeout time.Duration `config:"maxtimeout"`
+
+	// BackoffStrategy indicates which backoff strategy will be used between
+	// each retry attempt.
+	BackoffStrategy config.Backoff `config:"backoff"`
+}
+
+func (p PolicySpec) policy() (*Policy, error) {
+	strategy, err := p.BackoffStrategy.Strategy()
+	if err != nil {
+		return nil, err
+	}
+	return NewPolicy(
+		Retries(p.Retries),
+		MaxRequestTimeout(p.MaxRequestTimeout),
+		BackoffStrategy(strategy),
+	), nil
+}
+
+// OverrideSpec defines per service or per service+procedure
+// Policies that will be applied in the PolicyProvider.
+type OverrideSpec struct {
+	// Service is a YARPC service name for an override.
+	Service string `config:"service"`
+
+	// Procedure is a YARPC procedure name for an override.
+	Procedure string `config:"procedure"`
+
+	// WithPolicy specifies the policy name to use for the override.
+	// It MUST reference an existing policy.
+	WithPolicy string `config:"with"`
+}
+
+// PolicyProviderSpec is a definition of how to create a retry
+// policy provider.
+type PolicyProviderSpec struct {
+	// NameToPolicies is a map of names to policy configs which
+	// can be referenced later.
+	NameToPolicies map[string]PolicySpec `config:"policies"`
+
+	// Default is the Default policy that will be used.
+	Default string `config:"default"`
+
+	// Overrides are custom overrides for retry policies.
+	Overrides []OverrideSpec `config:"overrides"`
+}
+
+// NewPolicyProvider creates a new policy provider that can be used in retry
+// middleware
+func NewPolicyProvider(src interface{}, opts ...mapdecode.Option) (*ProcedurePolicyProvider, error) {
+	var err error
```

Comment:
nit: define variables as close to as where you need them.

In this case, the first if statement should inline its `err` declaration. You should declare this `var` next to `nameToPolicy`. Better, just do,

```go
nameToPolicy, err := spec.getPolicies(opts...)
if err != nil {
  return nil, err
}

return spec.getPolicyProvider(nameToPolicy)
```